### ISSUE TITLE
First-class support for Nextcloud

### ISSUE THREAD

**Issue by pemocarlo** [the-djmaze/snappymail#96]
Firt of all, thank you for this rewrite of the Rainloop app, it is really needed.

After reading the documentation, I saw that you removed support for all 3rd party apps, to ensure GDPR compliance.
But I wanted to know if you would be open to support Nextcloud - as an open source project where you are guaranteed to have control of your data, this should make sense.

Thanks a lot for your consideration

**Is your feature request related to a problem? Please describe.**
I have been using Rainloop with Nextcloud. In my opinion, it is the best email option available. Nevertheless, the integration between the two is currently less than perfect.

**Describe the solution you'd like**
Have a better integration with Nextcloud

**Describe alternatives you've considered**
Modify the Rainloop app to improve this, but the code is a bit outdated. I would prefer to work with this fork :)

**Additional context**
There is currently a Rainloop app for Nextcloud: https://github.com/pierre-alain-b/rainloop-nextcloud
Adding attachments from Nextcloud: https://github.com/RainLoop/rainloop-webmail/issues/2023

**Comment by the-djmaze** [comment]
> Firt of all, thank you for this rewrite of the Rainloop app, it is really needed.
Thanks for the heads up

> **Is your feature request related to a problem? Please describe.**
> I have been using Rainloop with Nextcloud. In my opinion, it is the best email option available. Nevertheless, the integration between the two is currently less than perfect.
> 
> **Describe the solution you'd like**
> Have a better integration with Nextcloud

Well, that's the problem. I have no time to make a better integration (maybe someone else does).
So why have a half-baked integration?
It's easier to have a standalone webmail and a link in Nextcloud to it.
Because when it's integrated, people will request features and someone has to create them, like:
* https://github.com/RainLoop/rainloop-webmail/issues/2107
* https://github.com/RainLoop/rainloop-webmail/issues/2079
* https://github.com/RainLoop/rainloop-webmail/issues/2023
* https://github.com/RainLoop/rainloop-webmail/issues/2013
* etc.

**Comment by Mer0me** [comment]
We're about to switch 600 users from Exchange/Outlook to Nextcloud/rainloop, but a snappymail Nextcloud app should be greater. Is it complicated to reproduce the rainloop-nextcloud app with SnappyMail ?

**Comment by carlosperezm** [comment]
@the-djmaze Thank you for your reply!

I completely agree with you that having a half-baked integration is not ideal. I guess that a Nextcloud integration fits better in a plugin.
I might give it a try, but I am kind of a beginner. Do you have some basic instructions on how to create a development environment for snappy mail?

**Comment by the-djmaze** [comment]
> I might give it a try, but I am kind of a beginner.

For that i've started a branch named OwnCloud
https://github.com/the-djmaze/snappymail/tree/OwnCloud
In there i made some changes to the source and a plugin.
https://github.com/the-djmaze/snappymail/commit/6e26c767c31b6edb51a6393155db13a3f20c66f9

> Do you have some basic instructions on how to create a development environment for snappy mail?

Read https://github.com/the-djmaze/snappymail/blob/master/CONTRIBUTING.md

**Comment by the-djmaze** [comment]
P.S. i've downloaded it and see that owncloud is different from nextcloud these days.

I will wrap up some basic code so you have a better start.

**Comment by carlosperezm** [comment]
Thanks for this!

> P.S. i've downloaded it and see that owncloud is different from nextcloud these days.
Yes, indeed, Nextcloud is a fork of Owncloud, which started back in 2016. I guess they still share most of the code base, but they will definitely continue diverging

Rainloop mail has some default options to add attachments to owncloud, dropbox, etc. I haven't tried snappy mail yet, but from what I've read, you removed all the related code.

In that case, do you think about adding something similar, in this case to Nextcloud, but better as a plugin?

**Comment by the-djmaze** [comment]
> but better as a plugin?

@carlosperezm i've modified and put the NextCloud code of pierre alain in the branch.
You can start with that to make the plugin and integration better.
The plugin is in /plugins/nextcloud

**Comment by Mer0me** [comment]
If I can make 3 wishes to developers of Snappymail/Nextcloud app(s) : 
- Create an official app to install Snappymail in Nextcloud as the rainloop app does
- Allow SnappyMail to save mail and attachments to Nextcloud, and to pick attachments from Nextcloud
- Add SnappyMail search into the Nextcloud search bar, and show last received mails in Nextcloud dashboard

That would be awesome !

**Comment by Mer0me** [comment]
I've forgotten a 4th wish : be able to insert .ics attachments in the Nextcloud calendar !

**Comment by scubamuc** [comment]
@the-djmaze 

nextcloud integration plans is great news! thank you for your fantastic work!

**Comment by the-djmaze** [comment]
The code is merged into master
https://github.com/the-djmaze/snappymail/tree/master/integrations/nextcloud/snappymail

You can use that to further develop de integration.
I can't because i don't have NextCloud nor OwnCloud

**Comment by the-djmaze** [comment]
I've installed NextCloud 22.1 locally to see if i can do something.
First i need to report something to them at HacherOne

**Comment by webagroprom** [comment]
Any news on nextcloud rainloop integration with add attachments feature? Thanks.

**Comment by Mer0me** [comment]
Work in progress here : we were able to add a button into the compose window and, using the [Nextcloud Filepicker project](https://github.com/eneiluj/nextcloud-webdav-filepicker), it opens a dialog box where we can select one or more files in Nextcloud and generate and copy the link(s) in the message body.
TO DO : authenticate the nextcloud user with oauth2 (auth is hardcoded for now)

Stay tuned !

**Comment by cm-schl** [comment]
Hi,

I'm sorry but am I right that at the moment the Nextcloud integration isn't officially available?

**Comment by scubamuc** [comment]
yip, that is correct. 
consider it work in progress... 
the community is doing excellent work, so you'll have to be patient if you can't contribute.

**Comment by scubamuc** [comment]
> And someone else also seems to be working on the file picker #307

great to see [](https://avatars.githubusercontent.com/u/386857?v=4)Avatar
Pierre-Alain on board !

**Comment by HeyVern** [comment]
I am very excited about a Snappymail integration with Nextcloud.  I'd be happy to help with testing.  Is this thread the best place to discuss the project?

**Comment by husim0** [comment]
I would be very happy too to test it :) Rainloop is very good and "well integrated" with Nextcloud, but Snappy seems far better and modern !

**Comment by DragoBarny** [comment]
Hello,
I'm sorry but I don't see the snappymail package in the nextcloud (24.0.3) apps library
I copied snappymal's git to nextcloud's app directory (old fashioned way) and the app appears (icon and config)
but if I click on the administration link I just have a 'null' message that is displayed

**Comment by beankylla** [comment]
What can a non-code do to help ? :)

**Comment by schengawegga** [comment]
> What can a non-code do to help ? :)



> Hello, I'm sorry but I don't see the snappymail package in the nextcloud (24.0.3) apps library I copied snappymal's git to nextcloud's app directory (old fashioned way) and the app appears (icon and config) but if I click on the administration link I just have a 'null' message that is displayed

I´ve got the same problem. Snappymail is not in the app library (NC 24.0.2) and after copying snappymail manually in the app directory and activating, only gives a "null" response on the app site and the admin site.

**Comment by the-djmaze** [comment]
I've added a plugin named `nextcloud`.
It should give people a start for integration plugins.
You can find it at https://github.com/the-djmaze/snappymail/tree/master/plugins/nextcloud

**Comment by the-djmaze** [comment]
I've changed the Nextcloud example code to be v18+ compatible and use better security.
Now the stored passwords use the SnappyMail Crypt class to properly encrypt the login password.

I could not test the code yet, as i have just setup a Nextcloud with error `This version of Nextcloud is not compatible with PHP>=8.2.`

**Comment by szaimen** [comment]
> I could not test the code yet, as i have just setup a Nextcloud with error `This version of Nextcloud is not compatible with PHP>=8.2.`

You could try https://github.com/nextcloud/all-in-one#nextcloud-all-in-one which shouldn't have this issue.

**Comment by the-djmaze** [comment]
The base code is working now.
![afbeelding](https://user-images.githubusercontent.com/3752035/195037494-f680e80c-4ce7-42a5-a01e-1a097e32105b.png)

**Comment by the-djmaze** [comment]
Feel free to try:

[snappymail-2.18.5-nextcloud.tar.gz](https://github.com/the-djmaze/snappymail/files/9753842/snappymail-2.18.5-nextcloud.tar.gz)

**Comment by pierre-alain-b** [comment]
Excellent, I will test soon!

**Comment by the-djmaze** [comment]
https://github.com/the-djmaze/snappymail/releases/tag/v2.18.6

**Comment by the-djmaze** [comment]
The embedded mode (without iframe) should work, but the CSS is a bit weird:
![afbeelding](https://user-images.githubusercontent.com/3752035/195544258-ecf58a44-30f7-4a0d-85f6-47ed568deacb.png)

vs original
![afbeelding](https://user-images.githubusercontent.com/3752035/195544363-663b59b7-f0cc-477a-90e6-a56d49932b8b.png)

**Comment by the-djmaze** [comment]
Most things are fixed now for `embed` mode.
Currently i found 1 annoying bug with it: `ServiceWorker Notifications not supported`

**Comment by the-djmaze** [comment]
@Mer0me see the bottom message menu option for next plugin release ;)
![afbeelding](https://user-images.githubusercontent.com/3752035/195592513-c9dd6ae0-73b3-46d4-8dfa-11431cb68cab.png)

**Comment by the-djmaze** [comment]
@Mer0me
> The killing feature of Nextcloud Mail is unified search

Ha! I've added mockup code for it in last commit.
Maybe someone will write it 😀

**Comment by jdaviescoates** [comment]
Wonderful to see so much progress here, thanks all!

Presumably this will eventually be published to the Nextcloud app store?

**Comment by pierre-alain-b** [comment]
Excellent indeed! Happy to redirect the rainloop nextcloud plugin to a snappymail one.
Let me know if I can be of any help.

**Comment by cm-schl** [comment]
Would be happy to help testing the Nextcloud integration 😃 - could you maybe give some info how to manually install snappymail and the Nextcloud plugin in a existing Nextcloud test environment?

**Comment by the-djmaze** [comment]
> I can eventually help set the plugin in the nextcloud app store but there is nothing complicated there

Well... https://github.com/nextcloud/appstore/issues/929

**Comment by the-djmaze** [comment]
> could you maybe give some info how to manually install

1. Download the snappymail-2.18.6-nextcloud.tar.gz
2. Extract it
3. Move/upload the extracted `snappymail` folder into you Nextcloud `/apps/` folder
4. Nextcloud -> Apps install the app
5. Nextcloud -> Settings -> Admin additional -> open the snappymail admin
6. Read https://github.com/the-djmaze/snappymail/wiki/Installation-instructions#admin-password

**Comment by pierre-alain-b** [comment]
> > I can eventually help set the plugin in the nextcloud app store but there is nothing complicated there
> 
> Well... i can't login [nextcloud/appstore#929](https://github.com/nextcloud/appstore/issues/929) So feel free to put it in the store.

it works here with FF 104... have you tried requesting a new password? I believe there was an update in the platform with a change of password needed.

**Comment by the-djmaze** [comment]
@Mer0me almost working :smile_cat: ...
![afbeelding](https://user-images.githubusercontent.com/3752035/195669847-dbc7d617-c618-4fe2-a4df-088ec5ea4ffd.png)

**Comment by the-djmaze** [comment]
@cm-schl i've just tested my unified search integration and it can search the same as the advanced search of mail app.

For example search `keyword=$label3` or `keyword:$label3` and it only shows messages with that tag/keyword/label.

So basically all works the sames.
The only problem is that i can't create direct links to messages yet.
This requires further test and merging of my `message-routing` branch, as discussed in #100

**Comment by the-djmaze** [comment]
I've updated the release with the new options and some redesign.

Please use this version and the new plugin for further testing.

App: https://github.com/the-djmaze/snappymail/releases/download/v2.18.6/snappymail-2.18.6-nextcloud.tar.gz

Plugin: https://snappymail.eu/repository/v2/plugins/nextcloud-2.2.tgz

**Comment by the-djmaze** [comment]
This last commit has a working WebDAV to PROPFIND the files/directories.
It should allow things like:
- Add Nextcloud attachments in message composer window
- Store message EML in selected folder (current hardcoded to `Emails`)
- Store message attachments in selected folder (current hardcoded to `Attachments`)

Now i need some time off of this and wait on you all if it works AND is someone is able to help with the integration.

There should be enough base code now to extend it all with your own wishes.

**Comment by relikd** [comment]
chiming in, haven't tested it yet. Currently I am using Rainloop but consider moving to SnappyMail. Thanks for the effort so far :) 👍  Can you check if the link to SnappyMail is visible in the mobile app? Because Rainloop does not show up, only external Links. But maybe this is the wrong place to ask and I should rather move the question to Nextcloud issues.

**Comment by the-djmaze** [comment]
@relikd 
> Thanks for the effort so far :) +1

Thank you!



> Can you check if the link to SnappyMail is visible in the mobile app?

No idea, but wouldn't it be better to use something like [FairEmail](https://email.faircode.eu/)

I do use SnappyMail and FairEmail on mobile for their own purposes.

**Comment by relikd** [comment]
I dont want to hijack this thread, so I keep it short. My use-case is a nextcould instance for a non-profit to allow collaboration. The email accounts are pre-configured and assigned to specific users. Each user can have 0-n email accounts. Giving away login/passwords would require to change the pw each time a user quits and the remaining users would need to update their stored pw.
But again, I dont want to hijack the thread. Just a feature wish: as an admin I would like to pre-configure accounts for others. Currently I do that with the Impersonate app.

**Comment by rainerzufall** [comment]
I am very happy to see that the nextcloud integration is happening. Thank you so much for your effort and time spent on this, the-djmaze!! 🙇
Just wanted to let you know, that I will be able to help test as soon it hits the nextcloud AppStore, as I am not tech savvy enough to run my own nextcloud instance and test at the current stage of development.

**Comment by hampoelz** [comment]
**Bug Report**

If the mail autologin credentials are entered immediately after the app installation before adding the server details in the admin panel, the mail app and admin panel cannot be accessed due to an internal server error.

<details>
  <summary>server log</summary>
  
  ```json
{
    "reqId": "Y0qzFwNr7Zdftc9y6lr6NQAAAAI",
    "level": 3,
    "time": "2022-10-15T13:18:16+00:00",
    "remoteAddr": "2001:871:20b:84d7:e815:ab92:d343:8d7",
    "user": "hampoelz",
    "app": "index",
    "method": "GET",
    "url": "/index.php/apps/snappymail/",
    "message": "ConnectionError[104]",
    "userAgent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/106.0.0.0 Safari/537.36 Edg/106.0.1370.42",
    "version": "25.0.0.17",
    "exception": {
        "Exception": "RainLoop\\Exceptions\\ClientException",
        "Message": "ConnectionError[104]",
        "Code": 104,
        "Trace": [
            {
                "file": "/home/webpages/hampoelz/Office/Nextcloud/apps/snappymail/app/snappymail/v/2.18.6/app/libraries/RainLoop/Actions/UserAuth.php",
                "line": 129,
                "function": "CheckMailConnection",
                "class": "RainLoop\\Actions",
                "type": "->",
                "args": [
                    {
                        "__class__": "RainLoop\\Model\\MainAccount"
                    },
                    true
                ]
            },
            {
                "file": "/home/webpages/hampoelz/Office/Nextcloud/apps/snappymail/lib/Util/SnappyMailHelper.php",
                "line": 47,
                "function": "LoginProcess",
                "class": "RainLoop\\Actions",
                "type": "->",
                "args": [
                    "example@example.org",
                    "example-pwd",
                    false
                ]
            },
            {
                "file": "/home/webpages/hampoelz/Office/Nextcloud/apps/snappymail/lib/Controller/PageController.php",
                "line": 86,
                "function": "startApp",
                "class": "OCA\\SnappyMail\\Util\\SnappyMailHelper",
                "type": "::",
                "args": []
            },
            {
                "file": "/home/webpages/hampoelz/Office/Nextcloud/apps/snappymail/lib/Controller/PageController.php",
                "line": 22,
                "function": "index_embed",
                "class": "OCA\\SnappyMail\\Controller\\PageController",
                "type": "::",
                "args": []
            },
            {
                "file": "/home/webpages/hampoelz/Office/Nextcloud/lib/private/AppFramework/Http/Dispatcher.php",
                "line": 225,
                "function": "index",
                "class": "OCA\\SnappyMail\\Controller\\PageController",
                "type": "->",
                "args": []
            },
            {
                "file": "/home/webpages/hampoelz/Office/Nextcloud/lib/private/AppFramework/Http/Dispatcher.php",
                "line": 133,
                "function": "executeController",
                "class": "OC\\AppFramework\\Http\\Dispatcher",
                "type": "->",
                "args": [
                    {
                        "__class__": "OCA\\SnappyMail\\Controller\\PageController"
                    },
                    "index"
                ]
            },
            {
                "file": "/home/webpages/hampoelz/Office/Nextcloud/lib/private/AppFramework/App.php",
                "line": 172,
                "function": "dispatch",
                "class": "OC\\AppFramework\\Http\\Dispatcher",
                "type": "->",
                "args": [
                    {
                        "__class__": "OCA\\SnappyMail\\Controller\\PageController"
                    },
                    "index"
                ]
            },
            {
                "file": "/home/webpages/hampoelz/Office/Nextcloud/lib/private/Route/Router.php",
                "line": 298,
                "function": "main",
                "class": "OC\\AppFramework\\App",
                "type": "::",
                "args": [
                    "PageController",
                    "index",
                    {
                        "__class__": "OC\\AppFramework\\DependencyInjection\\DIContainer"
                    },
                    [
                        "snappymail.page.index"
                    ]
                ]
            },
            {
                "file": "/home/webpages/hampoelz/Office/Nextcloud/lib/base.php",
                "line": 1047,
                "function": "match",
                "class": "OC\\Route\\Router",
                "type": "->",
                "args": [
                    "/apps/snappymail/"
                ]
            },
            {
                "file": "/home/webpages/hampoelz/Office/Nextcloud/index.php",
                "line": 36,
                "function": "handleRequest",
                "class": "OC",
                "type": "::",
                "args": []
            }
        ],
        "File": "/home/webpages/hampoelz/Office/Nextcloud/apps/snappymail/app/snappymail/v/2.18.6/app/libraries/RainLoop/Actions/UserAuth.php",
        "Line": 441,
        "Previous": {
            "Exception": "MailSo\\Net\\Exceptions\\SocketCanNotConnectToHostException",
            "Message": "Can't connect to host \"tcp://localhost:143\"",
            "Code": 0,
            "Trace": [
                {
                    "file": "/home/webpages/hampoelz/Office/Nextcloud/apps/snappymail/app/snappymail/v/2.18.6/app/libraries/MailSo/Imap/ImapClient.php",
                    "line": 91,
                    "function": "Connect",
                    "class": "MailSo\\Net\\NetClient",
                    "type": "->",
                    "args": [
                        {
                            "__class__": "MailSo\\Net\\ConnectSettings",
                            "host": "localhost",
                            "port": 143,
                            "type": 0,
                            "ssl": {
                                "0": "And 1 more entries, set log level to debug to see all entries",
                                "verify_peer": false,
                                "verify_peer_name": false,
                                "allow_self_signed": true,
                                "SNI_enabled": true,
                                "disable_compression": true
                            }
                        }
                    ]
                },
                {
                    "file": "/home/webpages/hampoelz/Office/Nextcloud/apps/snappymail/app/snappymail/v/2.18.6/app/libraries/RainLoop/Model/Account.php",
                    "line": 226,
                    "function": "Connect",
                    "class": "MailSo\\Imap\\ImapClient",
                    "type": "->",
                    "args": [
                        {
                            "__class__": "MailSo\\Net\\ConnectSettings",
                            "host": "localhost",
                            "port": 143,
                            "type": 0,
                            "ssl": {
                                "0": "And 1 more entries, set log level to debug to see all entries",
                                "verify_peer": false,
                                "verify_peer_name": false,
                                "allow_self_signed": true,
                                "SNI_enabled": true,
                                "disable_compression": true
                            }
                        }
                    ]
                },
                {
                    "file": "/home/webpages/hampoelz/Office/Nextcloud/apps/snappymail/app/snappymail/v/2.18.6/app/libraries/RainLoop/Actions/UserAuth.php",
                    "line": 437,
                    "function": "ImapConnectAndLoginHelper",
                    "class": "RainLoop\\Model\\Account",
                    "type": "->",
                    "args": [
                        {
                            "__class__": "RainLoop\\Plugins\\Manager"
                        },
                        {
                            "__class__": "MailSo\\Mail\\MailClient"
                        },
                        {
                            "__class__": "RainLoop\\Config\\Application"
                        }
                    ]
                },
                {
                    "file": "/home/webpages/hampoelz/Office/Nextcloud/apps/snappymail/app/snappymail/v/2.18.6/app/libraries/RainLoop/Actions/UserAuth.php",
                    "line": 129,
                    "function": "CheckMailConnection",
                    "class": "RainLoop\\Actions",
                    "type": "->",
                    "args": [
                        {
                            "__class__": "RainLoop\\Model\\MainAccount"
                        },
                        true
                    ]
                },
                {
                    "file": "/home/webpages/hampoelz/Office/Nextcloud/apps/snappymail/lib/Util/SnappyMailHelper.php",
                    "line": 47,
                    "function": "LoginProcess",
                    "class": "RainLoop\\Actions",
                    "type": "->",
                    "args": [
                        "example@example.org",
                        "example-pwd",
                        false
                    ]
                },
                {
                    "file": "/home/webpages/hampoelz/Office/Nextcloud/apps/snappymail/lib/Controller/PageController.php",
                    "line": 86,
                    "function": "startApp",
                    "class": "OCA\\SnappyMail\\Util\\SnappyMailHelper",
                    "type": "::",
                    "args": []
                },
                {
                    "file": "/home/webpages/hampoelz/Office/Nextcloud/apps/snappymail/lib/Controller/PageController.php",
                    "line": 22,
                    "function": "index_embed",
                    "class": "OCA\\SnappyMail\\Controller\\PageController",
                    "type": "::",
                    "args": []
                },
                {
                    "file": "/home/webpages/hampoelz/Office/Nextcloud/lib/private/AppFramework/Http/Dispatcher.php",
                    "line": 225,
                    "function": "index",
                    "class": "OCA\\SnappyMail\\Controller\\PageController",
                    "type": "->",
                    "args": []
                },
                {
                    "file": "/home/webpages/hampoelz/Office/Nextcloud/lib/private/AppFramework/Http/Dispatcher.php",
                    "line": 133,
                    "function": "executeController",
                    "class": "OC\\AppFramework\\Http\\Dispatcher",
                    "type": "->",
                    "args": [
                        {
                            "__class__": "OCA\\SnappyMail\\Controller\\PageController"
                        },
                        "index"
                    ]
                },
                {
                    "file": "/home/webpages/hampoelz/Office/Nextcloud/lib/private/AppFramework/App.php",
                    "line": 172,
                    "function": "dispatch",
                    "class": "OC\\AppFramework\\Http\\Dispatcher",
                    "type": "->",
                    "args": [
                        {
                            "__class__": "OCA\\SnappyMail\\Controller\\PageController"
                        },
                        "index"
                    ]
                },
                {
                    "file": "/home/webpages/hampoelz/Office/Nextcloud/lib/private/Route/Router.php",
                    "line": 298,
                    "function": "main",
                    "class": "OC\\AppFramework\\App",
                    "type": "::",
                    "args": [
                        "PageController",
                        "index",
                        {
                            "__class__": "OC\\AppFramework\\DependencyInjection\\DIContainer"
                        },
                        [
                            "snappymail.page.index"
                        ]
                    ]
                },
                {
                    "file": "/home/webpages/hampoelz/Office/Nextcloud/lib/base.php",
                    "line": 1047,
                    "function": "match",
                    "class": "OC\\Route\\Router",
                    "type": "->",
                    "args": [
                        "/apps/snappymail/"
                    ]
                },
                {
                    "file": "/home/webpages/hampoelz/Office/Nextcloud/index.php",
                    "line": 36,
                    "function": "handleRequest",
                    "class": "OC",
                    "type": "::",
                    "args": []
                }
            ],
            "File": "/home/webpages/hampoelz/Office/Nextcloud/apps/snappymail/app/snappymail/v/2.18.6/app/libraries/MailSo/Net/NetClient.php",
            "Line": 194
        },
        "CustomMessage": "--"
    },
    "id": "634ab47225aa0"
}
  ```
  
</details>

**Comment by hampoelz** [comment]
Once the embed mode without iframe works without problems - and thus the css variables of nextcloud can be accessed - I would help create a Nextcloud 25 theme, to make the app experience as native as possible.

For now, I played around a bit and added NC25's blurred background, buttons and scrollbars:

<details>
  <summary>CSS</summary>

  ```css
/*
 * blurred background
 */

html {
    background-color: rgba(255,255,255, .8);
    backdrop-filter: blur(25px);
    -webkit-backdrop-filter: blur(25px);
}

body {
    margin: 10px;
    height: calc(100% - 20px) !important;
    width: calc(100% - 20px) !important;
}

/*
 * bill buttons
 */

.messageList {
    margin: 0 5px;
}

a.selectable {
    margin: 2px;
    height: 38px !important;
    line-height: 38px !important;
    border-radius: 100px;
}

a.selectable::after {
   display: block;
   margin: 0 !important;
   top: 50%;
   transform: translateY(-50%);
}

.btn {
    border: initial !important;
}

/*
 * scrollbar
 */

::-webkit-scrollbar {
    width: 12px;
    height: 12px;
}

::-webkit-scrollbar-thumb {
    background: #bec2c9;
    border-radius: 10px;
    border: 2px solid transparent;
    background-clip: content-box;
}

::-webkit-scrollbar-track-piece {
    background-color: transparent;
}
  ```
  
</details>

The dark mode is quick and dirty so far, but it works well (mails are also automatically themed dark):

<details>
  <summary>CSS</summary>
  
  ```css
/*
 * dark mode
 */
@media (prefers-color-scheme: dark) {
    /*
     * blurred background
     */
    html {
        background-color: rgba(23,23,23, .8);
    }
    /**
     * Prevent native dark elements
     */
    :root {
        color-scheme: only light;
    }

    /**
     * Invert everything except images (and add a slight blue tint with the sepia() filter)
     */
    body, body img, #V-PopupsCompose, #V-PopupsCompose img {
        filter: invert(95%) sepia(15%) hue-rotate(180deg);
    }

    /**
     * Round the edges of images to make white images look better on dark backgrounds
     */
    img {
        border-radius: 4px;
    }

    /**
     * Remove shadows as they don't look good
     */
    *, *:before, *:after {
        box-shadow: none !important;
        text-shadow: none !important;
    }
}
  ```
  
</details>

Screenshots:

![light](https://user-images.githubusercontent.com/28979769/195998648-4f95f730-79d1-4f7c-a293-681f442b84ad.png)

![dark](https://user-images.githubusercontent.com/28979769/195998654-28d2e6b7-3f8a-413d-809c-0844049061e1.png)

**Comment by the-djmaze** [comment]
@hampoelz nice!
You might also take a look at https://github.com/the-djmaze/snappymail/blob/master/snappymail/v/0.0.0/themes/SquaresDark/styles.css
And put that into your `@media (prefers-color-scheme: dark) {`

**Comment by the-djmaze** [comment]
Wow, to get registered at Nextcloud app store is really a total mess!
Took me days to get something rolling (each browser different login errors).
https://github.com/nextcloud/app-certificate-requests/pull/549

**Comment by daffydock** [comment]
> chiming in, haven't tested it yet. Currently I am using Rainloop but consider moving to SnappyMail. Thanks for the effort so far :) +1 Can you check if the link to SnappyMail is visible in the mobile app? Because Rainloop does not show up, only external Links. But maybe this is the wrong place to ask and I should rather move the question to Nextcloud issues.

...Or you could try K9 for mobile. It is really leaping forward via recent updates, ever since the Thunderbird Project brought them under their wing.

**Comment by hampoelz** [comment]
> @hampoelz nice!
> You might also take a look at https://github.com/the-djmaze/snappymail/blob/master/snappymail/v/0.0.0/themes/SquaresDark/styles.css
> And put that into your `@media (prefers-color-scheme: dark) {`

Yeah, I've seen the two existing dark themes, but since I'm not entirely happy with them and don't have much time to play around with the colors right now, I've tweaked the hues a bit with invert, sepia and hue-rotate.

But as I said, I plan to create a snappymail-theme with the nextcloud colors later, so it should be possible to theme snappymail with nextcloud-themes, like the other nextcloud apps.

**Comment by daffydock** [comment]
Quick question, on your Release page for 2.18.6 you mention:

> NOTE: data is stored in {your-nextcloud-data-directory}/appdata_snappymail/_data_/ folder
>So it's NOT in your /{path-to-nextcloud}/apps/snappymail/

Is that only for now? Or, is this the final location for data moving forward? Either way is fine to me. I am just curious if the location is only temporary or not.

Cheers.

**Comment by the-djmaze** [comment]
@daffydock for now this is the best approach (i'm not a Nextcloud expert, just experienced Google user).
Because the default `data` location of the RainLoop app was in public.


I also made a first release candidate available https://apps.nextcloud.com/apps/snappymail

**Comment by daffydock** [comment]
> @daffydock for now this is the best approach (i'm not a Nextcloud expert, just experienced Google user). Because the default `data` location of the RainLoop app was in public.
> 
> I also made a first release candidate available https://apps.nextcloud.com/apps/snappymail

Funny, I just came back to update --which you already did-- that Snappymail is now available on the App store, which was not, when I originally posted my last comment. I am going to download it from there, if for sheer laziness. See how it works!

**Comment by SelfRef** [comment]
I saw this great news and installed SnappyMail on my server (via command `occ app:install snappymail --allow-unstable`)

In my case there is error on loading, only spinner shows up and in the console there is this error:
Firefox:
![image](https://user-images.githubusercontent.com/69125050/196286095-d8d4f942-7887-4f20-ac92-e23db451083f.png)
Chromium:
![image](https://user-images.githubusercontent.com/69125050/196286114-2a28af10-d228-4d28-9ddb-041025c16d0a.png)

RainLoop is working correctly on same server.

**Comment by webagroprom** [comment]
Im install snappy mail for nextcloud : 
`sudo -u nginx php occ app:install snappymail --allow-unstable`

snappymail 2.19.0-rc1 installed
snappymail enabled

change passwd: nextcloud/data/appdata_snappymail/_data_/_default_/admin_password.txt

Go admin panel and setting, open link admin panel snappy mail >> insert login: admin and passwd  from admin_password.txt

**Authentication failed**

![1666051953166](https://user-images.githubusercontent.com/30408997/196306484-5bc74dce-5e05-459d-ac31-e15f5aba2366.jpg)

how to fix ?

**Comment by the-djmaze** [comment]
@SelfRef the Nextcloud log should have information why you get a 500 error.

@webagroprom i've improved the Nextcloud admin settings to show the initial password and an option to enable/disable debug mode.

Should i make an RC2 or can you both test/debug without?

**Comment by the-djmaze** [comment]
I've released RC2 on Nextcloud because the short old info.xml terrible.
Now the app shows much better there.

https://apps.nextcloud.com/apps/snappymail

P.S. next release will have NC25 as it seems the info.xml had `<nextcloud min-version="20" max-version="24" />`

**Comment by Mer0me** [comment]
The app is not available on the app store yet ?

**Comment by the-djmaze** [comment]
@Mer0me currently it should be for NC20 - NC24 and not NC25 in an `unstable` state

**Comment by Mer0me** [comment]
I have 2 instances of Nextcloud 24.0.6, and the app is not visible on the app store when searching "snappy"

**Comment by the-djmaze** [comment]
@Mer0me strange, as `https://apps.nextcloud.com/api/v1/apps.json` does contain it.
Maybe Nextcloud caches it for some time?

**Comment by szaimen** [comment]
> I have 2 instances of Nextcloud 24.0.6, and the app is not visible on the app store when searching "snappy"

you need to set the channel to beta in order to see the release candidate on your apps page in Nextcloud

**Comment by the-djmaze** [comment]
Because i couldn't get the 'unstable' installed either, i've made a stable release on https://apps.nextcloud.com/apps/snappymail
That might work.

**Comment by Mer0me** [comment]
Still not visible for me.

**Comment by szaimen** [comment]
> Still not visible for me.

this is likely caching then

**Comment by the-djmaze** [comment]
@Mer0me i've got success now!
![afbeelding](https://user-images.githubusercontent.com/3752035/196407037-3856a066-0e3e-4add-ae43-2d082e76b775.png)

**Comment by Mer0me** [comment]
Yes ! Me too.

**Comment by relikd** [comment]
```
Exception: Class "PharData" not found in file '/var/snap/nextcloud/31222/nextcloud/extra-apps/snappymail/app/snappymail/v/2.19.0/app/libraries/snappymail/repository.php' line 269
```

**Comment by the-djmaze** [comment]
@relikd https://github.com/the-djmaze/snappymail/issues/392

I will built in a check for this.

**Comment by relikd** [comment]
Interesting. On my maschine, Phar is installed (`php -m` lists it). But either snap is using a different php version (in which case this is an issue for all snap users) or I dont know. Any way around this? I know if I copy the plugins manually, I will forget to update them regularly.

... the snap [config](https://github.com/nextcloud-snap/nextcloud-snap/blob/master/src/php/config/php.ini) even has a Phar option, maybe I will try that first...

**Comment by szaimen** [comment]
phar is not enabled on snap IIRC: https://github.com/nextcloud-snap/nextcloud-snap/blob/dd887feb8202597984e19cea132ababc61700c60/snap/snapcraft.yaml#L212

**Comment by Mer0me** [comment]
OK, it works now on my instance. Very good job which force me to reconsider the webmail option...
Saving .eml or attachments to Nextcloud and take an attachment from Nextcloud is not implemented yet ?

**Comment by Magneticdud** [comment]
Great work! Replaced rainloop without effort, and much faster than native nextcloud mail

For saving, it doesn't save the .eml file; for attachments it creates the Attachments folder (or different, in changed in the plugin setting, i guess), but then doesn't save the file

**Comment by szaimen** [comment]
Regardung this setting: Improvement idea: let each user choose an attachment folder as I know of cases where each user needs to have a different folder

**Comment by daffydock** [comment]
Ha, @relikd beat me to the punch. Earlier, I was able to install Snappymail via console, version 2.19.0r RC1, and was running just fine, sans plugins, as they did not seem to work. Which was fine, main functionality was there. Noticed an update available and from console, got the same error as them, with version 2.19.  

"Error: App with id snappymail has invalid signature"

**Comment by the-djmaze** [comment]
> For saving, it doesn't save the .eml file; for attachments it creates the Attachments

I will look into this issue

> Improvement idea: let each user choose an attachment folder in the user settings as I know of cases where each user needs to have a different folder

Actually there is a webdav.js that works. The idea is to create a popup view where you can select/create folder to put them in.

> Upon uninstalling and re-installing snappymail, I get this error from Nextcloud UI: "App with id snappymail has invalid signature"

Correct! I uploaded a new version with same name to see what will happen.
As it seems it will override the existing version!
This is something i need to watch out for in future that Nextcloud website doesn't warn or error for existing version and that it can take some time before the new signature will kick in.
I will fix this as well.

**Comment by the-djmaze** [comment]
@Mer0me i've fixed the plugin for you.
v2.4 should work and it has some draft code for a directory/files selector window.

**Comment by the-djmaze** [comment]
> Great work! Replaced rainloop without effort, and much faster than native nextcloud mail

Great to hear @Magneticdud !

**Comment by webagroprom** [comment]
> @SelfRef the Nextcloud log should have information why you get a 500 error.
> 
> @webagroprom i've improved the Nextcloud admin settings to show the initial password and an option to enable/disable debug mode.

1. Im update snappy mail ( remove or install )
```
sudo -u nginx php occ app:remove snappymail
sudo -u nginx php occ app:install snappymail --allow-unstable 
```
2. Enable Debug mod in admin panel
3. Cope login and pass from admin panel 
4. Open link - admin panel snappy mail

**admin panel**

![1666115623780](https://user-images.githubusercontent.com/30408997/196507681-2293bb5b-4061-4fd4-8692-aaf2adeaa81d.jpg)


**Constant Loading:**

![1666114950707](https://user-images.githubusercontent.com/30408997/196506148-b3d450f5-f475-480a-9ab1-33a375c54ef5.jpg)

Browser: safari, firefox
Nextcloud: 24.0.6 stable
**Log nextcloud is empty**

**Comment by SelfRef** [comment]
Indeed, removing SnappyMail completely and installing again solved the issue. Thanks @webagroprom.

**Comment by webagroprom** [comment]
> > @SelfRef the Nextcloud log should have information why you get a 500 error.
> > @webagroprom i've improved the Nextcloud admin settings to show the initial password and an option to enable/disable debug mode.
> 
> 1. Im update snappy mail ( remove or install )
> 
> ```
> sudo -u nginx php occ app:remove snappymail
> sudo -u nginx php occ app:install snappymail --allow-unstable 
> ```
> 
>  2. Enable Debug mod in admin panel
> 3. Cope login and pass from admin panel
> 4. Open link - admin panel snappy mail
> 
> **admin panel**
> 
> ![1666115623780](https://user-images.githubusercontent.com/30408997/196507681-2293bb5b-4061-4fd4-8692-aaf2adeaa81d.jpg)
> 
> **Constant Loading:**
> 
> ![1666114950707](https://user-images.githubusercontent.com/30408997/196506148-b3d450f5-f475-480a-9ab1-33a375c54ef5.jpg)
> 
> Browser: safari, firefox Nextcloud: 24.0.6 stable **Log nextcloud is empty** Rainloop is enable Snappy mail is enable
> 
> Im try change password on the example: 12345W and drop cache browser = result the same
> 
> After changing the password from the admin panel, I checked the default configuration file - the password data did not change:
> 
> `/nextcloud/data/appdata_snappymail/_data_/_default_/configs/application.ini`
> 
> ```
> ; Login and password for web admin panel
> admin_login = "admin"
> admin_password = "$2y$10$5gfr3EyNw4QdgZuPVwmMj.4zRXkkvUVbBHTv8YprEf/ajvO2NUGhn"
> admin_totp = ""
> ```
> 
> **Solution**
> 
> Remove app snappy mail or folder:
> 
> ```
> sudo -u nginx php occ app:remove snappymail
> rm -Rf /nextcloud/data/appdata_snappymail/
> ```
> 
> **Result** Log in correctly

**When changing the password to the snappy mail admin panel through the admin panel, the authorization data disappears:**

![1666123466038](https://user-images.githubusercontent.com/30408997/196532691-4e657cd9-16db-47d2-b81d-60cd1c89f5e9.jpg)

When you try to log into the admin panel with a new password:
**Constant loading**

![1666114950707](https://user-images.githubusercontent.com/30408997/196532520-1c629f17-59f1-4084-8a61-e775efd9c504.jpg)

**Comment by Mer0me** [comment]
> @Mer0me i've fixed the plugin for you. v2.4 should work and it has some draft code for a directory/files selector window.

Ok, saving .eml works now. Nice work !

**Comment by the-djmaze** [comment]
@webagroprom and @SelfRef i've figured it out!

When debug mode is on, SnappyMail uses the `Content-Security-Policy-Report-Only` header.
However, Nextcloud already set the `Content-Security-Policy` so there are two conflicting CSP headers.
Fixed this by removing the CSP header of Nextcloud.

**Comment by the-djmaze** [comment]
Release 2.19.0.1 and 2.19.01 not possible and 2.19.0-1 gets marked as unstable.
So i've made it 2.19.1

**You must manually update extension/plugin to 2.5!**

It includes:
* PharData workaround (just try it)
* Plugin has basic Nextcloud folder selector (no sub-tree yet) to save attachments/email
* Fix for Content-Security-Policy
* Some other cleanups and improvements

I shall close this ticket and any further issues/features can be done in new issues.
Because the current release already has much more integration then the RainLoop version.
I've also opened the Discussions section so you all can talk about this there as well
https://github.com/the-djmaze/snappymail/discussions

**Comment by jdaviescoates** [comment]
Just installed this. Once I'd set-up my domains in the admin settings it works great, many thanks!

**Comment by relikd** [comment]
> * PharData workaround (just try it)

works like a charm (y)

**Comment by beankylla** [comment]
thanks a lot @the-djmaze works like a charrmmm :)

**Comment by schengawegga** [comment]
Yesterday i installed snappymail on my nextcloud 24.0.5 via the app menu. It worked without any problems. Thanks you @the-djmaze for your amazing support.

**Comment by jiriks74** [comment]
Well, this looks promising!

<details>

<summary>Rant about old Rainloop or weird plugin setup in Roundcube</summary>

Since Rainloop is currently old (and probably unsecure) pain, I've pluginned the hell out of Roundcube to get multiple emails. I made it work well, but as it's a plugin it does create some weird behaviour like: you can setup special folders like Inbox for every account separately but you cannot setup Archive folder separately as that is a plugin. So you have to have the archive folder under the same name for all accounts (which is pain because of Gmail - in Czech it's called "Všechny zprávy" not Archive and it's not an imap folder, urghh...). But it kinda works

</details>

I'll sure take a look into this plugin as it's what I wanted in the begining. Simple, multiemail webapp that's easy to run with nextcloud (Roundcube doesn't run with Nextcloud at all...). Looks way to promising to be true actually!

I've seen that the Nextcloud integration is WIP, but I'll take that and thank you for it. 

This has just made my day and I haven't even tried it yet LOL
